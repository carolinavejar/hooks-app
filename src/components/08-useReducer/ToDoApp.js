import React, { useReducer } from 'react';
import '../css/index.css';
import { todoReducer } from './todoReducer';

export const TodoApp = () => {

    const initialState = [{
        id: new Date().getTime(),
        desc: 'Aprender React',
        done: false
    }]

    const [ todos, dispatch ] = useReducer(todoReducer, initialState);

    const handleSubmit = (e) => {
        e.preventDefault();
        const newTodo = {
            id: new Date().getTime(),
            desc: 'Nueva tarea',
            done: false
        }

        const action = {
            type: 'add',
            payload: newTodo
        };
        dispatch(action);
        console.log(`Nueva tarea`);
    };

    
    console.log(todos);
    return (
        <div id = "todoList">
            <h1>To Do App ( {todos.length} )</h1>
            <hr />

            <div className = "row">
                <div className="col-7">
                    
                    <ul>
                        {
                            todos.map( (todo, i) => (
                                <li
                                    key = { todo.id }
                                    className = "list-group-item"
                                >
                                    <small>{ i + 1 }. {todo.desc}</small>

                                    <button 
                                        className="btn btn-outline-primary ml-3" 
                                        id = "btnBorrar"
                                    >
                                        Borrar
                                    </button>
                                </li>
                            ))
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4>Agregar TO DO</h4>
                    <hr />
                    <form onSubmit = { handleSubmit }>
                        <div className = "d-grid gap-2">
                            <input
                                className = "form-control"
                                type = "text"
                                name = "decription"
                                placeholder = "Aprender ..."
                                autoComplete = "off"
                            />
                            <button 
                                type = "submit"
                                className="btn btn-outline-primary btn-block mt-1" 
                            >
                                Agregar
                            </button>
                        </div>
                        
                    </form>
                </div>
            </div>
        </div>
    )
}
