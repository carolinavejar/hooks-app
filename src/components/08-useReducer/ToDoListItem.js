import React from 'react'

export const ToDoListItem = ({todo, i, handleDelete, handleToggle}) => {
    return (
        <li
            key = { todo.id }
            className = "list-group-item"
        >
            <small
                className = { `${ todo.done &&  'complete' }`}
                onClick = { () => {  handleToggle(todo.id) } }
            >
                    { i + 1 }. {todo.desc}
            </small>

            <button 
                className="btn btn-outline-primary ml-3 btnBorrar" 
                id = { todo.id }
                onClick = { handleDelete }
            >
                Borrar
            </button>
        </li>
    )
}
