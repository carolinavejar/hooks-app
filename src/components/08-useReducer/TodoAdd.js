import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAdd }) => {
    const [{ description }, handleInputChange, reset] = useForm({
        description: '',
    }); 

    const handleSubmit = (e) => {
        e.preventDefault();
        if(description.trim().length <= 0){
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAdd(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar TO DO</h4>
            <hr />
            <form onSubmit = { handleSubmit }>
                <div className = "d-grid gap-2">
                    <input
                        className = "form-control"
                        type = "text"
                        name = "description"
                        placeholder = "Aprender ..."
                        autoComplete = "off"
                        value = { description }
                        onChange = { handleInputChange }
                    />
                    <button 
                        type = "submit"
                        className="btn btn-outline-primary btn-block mt-1" 
                    >
                        Agregar
                    </button>
                </div>
                
            </form>
        </>
    )
}
