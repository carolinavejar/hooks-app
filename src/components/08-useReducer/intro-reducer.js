const initialState = [ {
    id: 1,
    todo: 'Comprar pan', 
    done: false
}];

const todoReducer = (state = initialState, action) => {
    if(action?.type == "add") {
        return [...state, action.payload]
    }
    return state;
};

let toDos = todoReducer();
// aÃ±adir nuevo todo 
const newTodo = {
    id: 2,
    todo: 'Comprar chocolate', 
    done: false
};

const addToDoAction = {
    type: 'add',
    // newTodo
    payload : newTodo // ->estandar
}

toDos = todoReducer(toDos, addToDoAction);

console.log(toDos);