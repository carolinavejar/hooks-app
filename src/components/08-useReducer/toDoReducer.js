export const todoReducer = ( state = [], action ) => {
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
    
        case 'remove': 
            // The filter() method creates a new array with all elements that pass the test implemented by the provided function
            return state.filter( state2 => state2.id != action.payload.id);

        case 'toggle' :
            return state.map(todo => 
               todo.id == action.payload ? { ...todo, done: !todo.done } : todo
            );
        case 'toggle-old': 
            return state.map(todo => {
                if( todo.id == action.payload) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }
            })
        default:
            return state;
    }
}