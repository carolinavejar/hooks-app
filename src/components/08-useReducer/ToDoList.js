import React from 'react'
import { ToDoListItem } from './ToDoListItem';

export const ToDoList = ({todos, handleDelete, handleToggle}) => {
    console.log(todos);
    return (
        
        <ul>
            {
                todos.map( (todo, i) => (
                    <ToDoListItem
                        key = { todo.id }
                        todo = { todo }
                        i = { i }
                        handleDelete = { handleDelete }
                        handleToggle = { handleToggle}
                    />
                ))
            }
        </ul>
        
    )
}
