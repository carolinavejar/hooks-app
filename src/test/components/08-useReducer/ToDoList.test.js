import { demoTodos } from "../../fixtures/demoTodos";
import { shallow } from 'enzyme';
// import Adapter from 'enzyme-adapter-react-16';
// No las toma de setupTests.js , así que las dejaré aquí por ahora
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer } from 'enzyme-to-json';

Enzyme.configure({ adapter: new Adapter() });
expect.addSnapshotSerializer( createSerializer ( { mode : 'deep' } ));

import { ToDoList } from "../../../components/08-useReducer/ToDoList";

describe('Pruebas en To Do List', () => {
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    const wrapper = shallow(
        <ToDoList
            todos={ demoTodos }
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
        />
    )
            
    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe tener x TodoListItem', () => {
        expect(wrapper.find('ToDoListItem').length).toBe(demoTodos.length);
        expect(wrapper.find('ToDoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function)); 
    });

})