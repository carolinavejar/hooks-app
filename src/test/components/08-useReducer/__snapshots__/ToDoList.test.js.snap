// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en To Do List debe mostrarse correctamente 1`] = `
<ul>
  <ToDoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={0}
    key="1"
    todo={
      Object {
        "desc": "Aprenderr react",
        "done": false,
        "id": 1,
      }
    }
  />
  <ToDoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;

exports[`Pruebas en To Do List debe tener x TodoListItem 1`] = `
<ul>
  <ToDoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={0}
    key="1"
    todo={
      Object {
        "desc": "Aprenderr react",
        "done": false,
        "id": 1,
      }
    }
  />
  <ToDoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
